src/third_party/starboard/wpe/brcm/arm/player/player_video_sink.cc

```diff
@@ -0,0 +1,63 @@
+// Copyright 2016 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "third_party/starboard/wpe/shared/player/player_video_sink.h"
+
+static void ElementAdded(GstElement* pipeline, GstElement* element, void* arg) {
+
+  bool pip_mode = (bool) arg;
+
+  if ((g_strrstr(GST_ELEMENT_NAME(element), "uridecodebin"))
+      || (g_strrstr(GST_ELEMENT_NAME(element), "decodebin"))) {
+    g_signal_connect(element, "element-added",
+        G_CALLBACK(&ElementAdded), arg);
+  }
+  if (g_strrstr(GST_ELEMENT_NAME(element), "brcmvideodecoder")) {
+    if (pip_mode) {
+      g_object_set(element, "pip", TRUE, NULL);
+    }
+  }
+  if (g_strrstr(GST_ELEMENT_NAME(element), "brcmaudiodecoder")) {
+    g_object_set(element, "use-primer", TRUE, NULL);
+    if (pip_mode) { // Do not start audio decoder for pip window
+      g_object_set(element, "prime", TRUE, NULL);
+    }
+  }
+}
+
+void AddVideoSink(GstElement* pipeline, bool pip_mode) {
+
+  g_signal_connect(pipeline, "element-added", G_CALLBACK(&ElementAdded),
+      (void*) pip_mode);
+
+#if defined(WESTEROS_SINK)
+
+  GstElement* video_sink = gst_element_factory_make("westerossink",
+      "videosink");
+  if (pip_mode) {
+    if (g_object_class_find_property(G_OBJECT_GET_CLASS(video_sink), "pip")) {
+      g_object_set(G_OBJECT(video_sink), "pip", TRUE, NULL);
+    }
+  }
+  g_object_set(pipeline, "video-sink", video_sink, NULL);
+
+#else
+
+  GstElement* video_sink = gst_element_factory_make("brcmvideosink",
+      "videosink");
+  g_object_set(pipeline, "video-sink", video_sink, NULL);
+
+#endif
```

Comment:
couple of issues here:
1. westeros sink is not Brcm speciic. I can be there on aml or intel. It's not platform specific and can be done in a common code
2. Can't we just - in the shared code - set the "pip property if it's there?
3. No need to explicitly set the sink in other case than westeros sink
4. Could we split ElementAdded for proper cases instead of having it a bag for any element added anywhere? Or use deep-element-added signal insead?

Reply:
Also look at the utility I added for NF (and WK) to find proper elements without relying on platform specific, hardcoded names: https://github.com/Metrological/netflix/blob/5.2.3/partner/dpi/gstreamer/GstServer.cpp#L245

Reply:
If all the above could be addressed then we're down to only "prime" and "use-primer" to be really not that generic (which we can still do in the common code based on if such properties are supported or not if we want)

Reply:
Following were few observations:
--If brcmvideosink is not set explicitly, PIP mode doesn't work as expected.
--Westeros sink supports PIP mode only on Broadcom platform. Not on RPI. Not sure about Amlogic.

Reply:
1. The autodetect attaches that sink anyway so it does not sound right to me.
2. It does not matter as long as the property check is there IMO.